File: typescript/src/model/types/wdl-primitive-type.ts
/** Primitive type nodes and primitive type names. */
import { WdlType, WdlTypeComponentType } from './wdl-type.js';

/** Models a primitive WDL type such as `Int`, `String`, or `File`. */
export class WdlPrimitiveType extends WdlType {
  /** Creates a primitive type from its primitive name and optional flag. */
  public constructor(
    private primitiveValueType: WdlPrimitiveType.Type = WdlPrimitiveType.Type.STRING,
    optional = false,
  ) {
    super(optional);
  }

  /** Returns the broad type family for this node. */
  public componentType(): WdlTypeComponentType {
    return WdlTypeComponentType.PRIMITIVE;
  }

  /** Returns the primitive type name represented by this node. */
  public primitiveType(): WdlPrimitiveType.Type {
    return this.primitiveValueType;
  }

  /** Sets the primitive type name represented by this node. */
  public setPrimitiveType(primitiveValueType: WdlPrimitiveType.Type): void {
    this.primitiveValueType = primitiveValueType;
  }
}

export namespace WdlPrimitiveType {
  /** Primitive type names supported by the WDL specification. */
  export enum Type {
    BOOLEAN = 'Boolean',
    INT = 'Int',
    FLOAT = 'Float',
    STRING = 'String',
    FILE = 'File',
    DIRECTORY = 'Directory',
    OBJECT = 'Object',
  }
}

File: java/src/main/java/com/myriad/wdl/model/types/WdlPrimitiveType.java
package com.myriad.wdl.model.types;

import java.util.Objects;

/**
 * Primitive WDL type with optional marker.
 *
 * <p>This models primitive type names such as {@code Int}, {@code String}, {@code File}, and
 * {@code Directory} from the specification's "Primitive Types" section.
 */
public final class WdlPrimitiveType extends WdlType {
  /** Supported primitive type names. */
  public enum Type {
    BOOLEAN("Boolean"),
    INT("Int"),
    FLOAT("Float"),
    STRING("String"),
    FILE("File"),
    DIRECTORY("Directory"),
    OBJECT("Object");

    private final String wdlName;

    private Type(String wdlName) {
      this.wdlName = wdlName;
    }

    public String toWdlString() {
      return wdlName;
    }
  }

  private Type primitiveValueType = Type.STRING;

  public WdlPrimitiveType() {
    super();
  }

  public WdlPrimitiveType(Type primitiveValueType, boolean optional) {
    super(optional);
    this.primitiveValueType = Objects.requireNonNull(primitiveValueType, "primitiveValueType");
  }

  @Override
  public ComponentType componentType() {
    return ComponentType.PRIMITIVE;
  }

  /** Returns the specific primitive type name represented by this node. */
  public Type primitiveType() {
    return primitiveValueType;
  }

  /** Sets the specific primitive type name represented by this node. */
  public void setPrimitiveType(Type primitiveValueType) {
    this.primitiveValueType = Objects.requireNonNull(primitiveValueType, "primitiveValueType");
  }
}
